platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+          LOGGER.debug("Fragment [{}] not ready with state [{}]", bundleName, bundleState);
+          unavailableBundles.add(bundle);
+        }
+      } else if (!BundleState.Active.equals(bundleState)) {
+        LOGGER.debug("Bundle [{}] not ready with state [{}]", bundleName, bundleState);
+        unavailableBundles.add(bundle);
+      }
+    }
+
+    if (!failedBundles.isEmpty()) {
+      printBundleInformation(LOGGER::error, failedBundles);
+      throw new SynchronisedInstallerException(
+          "Bundles failed to start ["
+              + failedBundles.stream().map(Bundle::getSymbolicName).collect(Collectors.joining(","))
+              + "].");
+    }
+
+    if (!unavailableBundles.isEmpty()) {
+      LOGGER.info(
+          "Waiting on the following bundles to reach Active state [{}]",
+          unavailableBundles
+              .stream()
+              .map(Bundle::getSymbolicName)
+              .collect(Collectors.joining(",")));
+    }
+
+    return unavailableBundles;
+  }
+
+  @VisibleForTesting
+  protected ServiceTracker getServiceTracker(BundleContext bundleContext, Filter filter) {
+    return new ServiceTracker<>(bundleContext, filter, null);
+  }
+
+  private void printBundleInformation(
+      BiConsumer<String, Object[]> logConsumer, Collection<Bundle> bundles) {
+    for (Bundle bundle : bundles) {
+      StringBuilder headerString = new StringBuilder("[ ");
+      Enumeration<String> keys = bundle.getHeaders().keys();
+
+      while (keys.hasMoreElements()) {
+        String key = keys.nextElement();
+        headerString.append(key).append("=").append(bundle.getHeaders().get(key)).append(", ");
+      }
+
+      headerString.append(" ]");
+      logConsumer.accept(
+          "\n\tBundle: {}_v{} | {}\n\tHeaders: {}",
+          new Object[] {
+            bundle.getSymbolicName(),
+            bundle.getVersion(),
+            BUNDLE_STATES.getOrDefault(bundle.getState(), "UNKNOWN"),
+            headerString
+          });
+    }
+  }
+
+  private Stream<Feature> getFeaturesFromNames(String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    Stream.Builder<Feature> transformedFeatures = Stream.builder();
+
+    for (String f : toList(feature, additionalFeatures)) {
+      try {
+        transformedFeatures.add(featuresService.getFeature(f));
+      } catch (Exception e) {
+        throw new SynchronisedInstallerException("Failed to retrieve feature [" + f + "]", e);
+      }
+    }
+
+    return transformedFeatures.build();
+  }
+
+  private Filter getServiceIdFilter(String servicePid) throws SynchronisedInstallerException {
+    try {
+      return bundleContext.createFilter(String.format(SERVICE_PID_FILTER, servicePid));
+    } catch (Exception e) {
+      throw new SynchronisedInstallerException("Failed to create service filter.", e);
+    }
+  }
+
+  private Dictionary<String, Object> toDic(Map<String, Object> props) {
```

Comment:
❓ Why not use the `DictionaryMap` class?

Reply:
We really need to pull these generic utility classes in our own 3rdparty artifact like apache.lang